pkg/registry/apis/provisioning/jobs/export/resources.go

```diff
@@ -28,28 +32,70 @@ func ExportResources(ctx context.Context, options provisioning.ExportJobOptions,
 			return fmt.Errorf("get client for %s: %w", kind.Resource, err)
 		}
 
-		if err := exportResource(ctx, options, client, repositoryResources, progress); err != nil {
+		// When requesting v2 (or v0) dashboards over the v1 api, we want to keep the original apiVersion if conversion fails
+		var upgrader versionUpgrader
+		if kind.GroupResource() == resources.DashboardResource.GroupResource() {
+			var v2client dynamic.ResourceInterface
+			upgrader = func(ctx context.Context, item *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+				failed, _, _ := unstructured.NestedBool(item.Object, "status", "conversion", "failed")
+				if failed {
+					storedVersion, _, _ := unstructured.NestedString(item.Object, "status", "conversion", "storedVersion")
+
+					// For v2 we need to request the original version
+					if strings.HasPrefix(storedVersion, "v2") {
+						if v2client == nil {
+							v2client, _, err = clients.ForResource(resources.DashboardResourceV2)
+							if err != nil {
+								return nil, err
+							}
+						}
+						return v2client.Get(ctx, item.GetName(), metav1.GetOptions{})
+					}
+
+					// For v0 we can simply fallback -- the full model is saved, but
+					if strings.HasPrefix(storedVersion, "v0") {
+						item.SetAPIVersion(fmt.Sprintf("%s/%s", kind.Group, storedVersion))
+						return item, nil
+					}
+
+					return nil, fmt.Errorf("unsupported dashboard version: %s", storedVersion)
+				}
+				return item, nil
+			}
+		}
+
+		if err := exportResource(ctx, options, client, upgrader, repositoryResources, progress); err != nil {
 			return fmt.Errorf("export %s: %w", kind.Resource, err)
 		}
 	}
 
 	return nil
 }
 
-func exportResource(ctx context.Context, options provisioning.ExportJobOptions, client dynamic.ResourceInterface, repositoryResources resources.RepositoryResources, progress jobs.JobProgressRecorder) error {
-	return resources.ForEach(ctx, client, func(item *unstructured.Unstructured) error {
-		fileName, err := repositoryResources.WriteResourceFileFromObject(ctx, item, resources.WriteOptions{
-			Path: options.Path,
-			Ref:  options.Branch,
-		})
-
+func exportResource(ctx context.Context,
+	options provisioning.ExportJobOptions,
+	client dynamic.ResourceInterface,
+	upgrader versionUpgrader,
```

Comment:
I don't see what we grain from passing this function instead of having the condition within this function.

Reply:
then you have to pass in either the client factory and create a new client each time -- also this should only apply for dashboard